File: Day49/Part1/dissonant_mirrors.py
from collections import deque

def solve_dissonant_mirrors(n, m, grid):
    # BFS State: (r, c, grid_state)
    # grid_state: 0 = Normal, 1 = Inverted
    
    queue = deque([(0, 0, 0, 0)]) # r, c, state, steps
    visited = set([(0, 0, 0)])
    
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    
    while queue:
        r, c, state, steps = queue.popleft()
        
        if r == n - 1 and c == m - 1:
            return steps
        
        for dr, dc in directions:
            nr, nc = r + dr, c + dc
            
            if 0 <= nr < n and 0 <= nc < m:
                # Determine value of next cell in current state
                cell_val = grid[nr][nc] ^ state
                
                new_state = state
                if cell_val == 1:
                    new_state = 1 - state
                    
                if (nr, nc, new_state) not in visited:
                    visited.add((nr, nc, new_state))
                    queue.append((nr, nc, new_state, steps + 1))
                    
    return -1

if __name__ == "__main__":
    Grid = [
        [0, 1],
        [1, 0]
    ]
    N = len(Grid)
    M = len(Grid[0])
    
    result = solve_dissonant_mirrors(N, M, Grid)
    print(f"Minimum Steps: {result}")

File: Day49/Part1/dissonant_mirrors.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <tuple>
#include <set>

using namespace std;

struct State {
    int r, c, grid_state, steps;
};

int solve_dissonant_mirrors(int n, int m, const vector<vector<int>>& grid) {
    // BFS State: {r, c, grid_state}
    // grid_state: 0 = Normal, 1 = Inverted
    
    queue<State> q;
    q.push({0, 0, 0, 0});
    
    // Visited array: visited[r][c][state]
    vector<vector<vector<bool>>> visited(n, vector<vector<bool>>(m, vector<bool>(2, false)));
    visited[0][0][0] = true;
    
    int dr[] = {0, 0, 1, -1};
    int dc[] = {1, -1, 0, 0};
    
    while (!q.empty()) {
        State current = q.front();
        q.pop();
        
        if (current.r == n - 1 && current.c == m - 1) {
            return current.steps;
        }
        
        for (int i = 0; i < 4; ++i) {
            int nr = current.r + dr[i];
            int nc = current.c + dc[i];
            
            if (nr >= 0 && nr < n && nc >= 0 && nc < m) {
                int cell_val = grid[nr][nc] ^ current.grid_state;
                int new_state = current.grid_state;
                
                if (cell_val == 1) {
                    new_state = 1 - current.grid_state;
                }
                
                if (!visited[nr][nc][new_state]) {
                    visited[nr][nc][new_state] = true;
                    q.push({nr, nc, new_state, current.steps + 1});
                }
            }
        }
    }
    
    return -1;
}

int main() {
    vector<vector<int>> grid = {
        {0, 1},
        {1, 0}
    };
    
    int n = grid.size();
    int m = grid[0].size();
    
    cout << "Minimum Steps: " << solve_dissonant_mirrors(n, m, grid) << endl;
    
    return 0;
}
